src/add.rs

```diff
@@ -0,0 +1,189 @@
+use core::mem::transmute;
+
+use float::Float;
+
+macro_rules! add {
+    ($intrinsic:ident: $ty:ty) => {
+        /// Returns `a + b`
+        #[cfg_attr(not(test), no_mangle)]
+        pub extern fn $intrinsic(a: $ty, b: $ty) -> $ty {
+            let bits =             (<$ty>::bits()) as <$ty as Float>::Int;
+            let significand_bits = (<$ty>::significand_bits()) as <$ty as Float>::Int;
+            let exponent_bits =    (bits - significand_bits - 1) as <$ty as Float>::Int;
+            let max_exponent =     ((1 << exponent_bits) - 1) as <$ty as Float>::Int;
+
+            let implicit_bit =     (1 << significand_bits) as <$ty as Float>::Int;
+            let significand_mask = (implicit_bit - 1) as <$ty as Float>::Int;
+            let sign_bit =         (1 << (significand_bits + exponent_bits)) as <$ty as Float>::Int;
+            let abs_mask =         (sign_bit - 1) as <$ty as Float>::Int;
+            let exponent_mask =    (abs_mask ^ significand_mask) as <$ty as Float>::Int;
+            let inf_rep =          (exponent_mask) as <$ty as Float>::Int;
+            let quiet_bit =        (implicit_bit >> 1) as <$ty as Float>::Int;
+            let qnan_rep =         (exponent_mask | quiet_bit) as <$ty as Float>::Int;
```

Comment:
You can probably get rid of all the `as <$ty as Float>::Int` casts here. Type inference will figure it out.